# Welcome to RADCamp
Over the last 15 years biogeographers have increasingly transitioned from investigating
phylogeographic patterns in space and time using datasets composed of one or only a handful
of markers to massive datasets containing thousands or tens of thousands of "anonymous"
nuclear loci generated using restriction site associated DNA sequencing (RAD-Seq). These
larger datasets provide more robust phylogenetic estimates, and can provide additional sources
of information such as evidence of historical introgression. The process of organizing and
making sense of the vast quantities of reads that come back off a sequencing instrument is
non-trivial, and of great consequence.

RADCamp workshops are designed to introduce ipyrad, a unified and self-contained RAD-Seq 
assembly and analysis framework, which emphasizes simplicity, performance, and reproducibility. 
In these workshops we proceed through all the steps necessary to assemble a typical RAD-Seq dataset. 
Additionally, we introduce the ipyrad 'analysis' API which provides a powerful, simple, and 
reproducible interface to several widely used methods for inferring phylogenetic relationships, 
population structure, and admixture. 

# What is RAD-Seq?

* A canonical introduction: [Andrews et al 2016](https://www.nature.com/articles/nrg.2015.28)
* A thorough review of reduced representation protocols: [Campbell et al 2018](https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.13038)

# Who should attend?
RADCamp workshops are normally geared toward practicing field biologists with little or no 
computational experience.

Workshop attendees will need to bring a laptop computer.

## Current event:
* [RADCamp2025-Brisbane](Brisbane2025/index.md) (June 2025)

## Future events:
* [RADCamp2026-LatinAmerica](LatinAmerica2026/index.md) (July 2026)

## Past events:
* [RADCamp2024-Phoenix](Phoenix2024/index.md)
* [RADCamp2024-SanFrancisco](SanFrancisco2024/index.md)
* [RADCamp2024-Chicago](Chicago2024/index.md)
* [RADCamp2023-Chicago](Chicago2023/index.md)
* [RADCamp2023-Kigali](Kigali2023/index.md)
* [RADCamp2023-NYC](NYC2023/index.md)
* [RADCamp2020-Marseille](Marseille2020/index.md)
* [RADCamp2020-NYC](NYC2020/index.md)
* [RADCamp2020-Lisbon](Lisbon2020/index.md)
* [RADCamp2019-NYC](NYC2019/index.md)
* [RADCamp2019-Yale](Yale2019/index.md)
* [RADCamp2019-IBS](IBS2019/index.md)
* [RADCamp2018-NYC](NYC2018/index.md)
* [RADCamp2018-AF-Biota (Brazil)](AF-Biota/index.md)

# Sponsorship Acknowledgments

[We acknowledge generous support from all our previous sponsors](funders.md)

# RADCamp tutorial contributors and instructors (over the years)
Isaac Overcast, Deren Eaton, Sandra Hoffberg, Natalia Bayona-Vasquez, Mariana
Vasconcellos, Laura Bertola, Josiah Kuja, Anhubab Kahn, Arianna Kuhn, Edgar
Benavides, Fransico Pina Martins

# References
* Andrews, K. R., Good, J. M., Miller, M. R., Luikart, G., & Hohenlohe, P. A. (2016). Harnessing the power of RADseq for ecological and evolutionary genomics. *Nature Reviews Genetics*, 17(2), 81.
* Campbell, E. O., Brunet, B. M., Dupuis, J. R., & Sperling, F. A. (2018). Would an RRS by any other name sound as RAD?. *Methods in Ecology and Evolution*, 9(9), 1920-1927.

# Acknowledgements
[PoreCamp](https://porecamp.github.io/) - Which inspired the design of this workshop and also of this site.
